Yosemite/Yosemite/SnapshotsProvider/FetchResultSnapshotsProvider.swift

```diff
@@ -325,3 +374,48 @@ private struct ObjectsDidChangeNotification {
         (notification.userInfo?[NSUpdatedObjectsKey] as? Set<NSManagedObject>) ?? Set()
     }
 }
+
+// MARK: - StorageManager Reset Handling
+
+@available(iOS 13.0, *)
+private extension FetchResultSnapshotsProvider {
+
+    /// Observe `StorageManagerDidResetStorage` notifications so that we can restart the
+    /// `fetchedResultsController` fetching.
+    ///
+    /// This routine is to fix a crash when `CoreDataManager.reset()` is called (e.g. during
+    /// user log out). If we don't do this, it looks like the `self.fetchedResultsController`
+    /// would still be pointing to an invalid `NSManagedObjectContext`. If that is not corrected,
+    /// we would get a crash like this when something changes in the database:
+    ///
+    /// ```
+    /// [error] error: Serious application error.  Exception was caught during Core Data change
+    /// processing.  This is usually a bug within an observer of
+    /// NSManagedObjectContextObjectsDidChangeNotification.  Object's persistent store is not
+    /// reachable from this NSManagedObjectContext's coordinator with userInfo (null)
+    /// ```
+    ///
+    /// This crash comes from `NSFetchedResultsController` itself and not from the observation
+    /// in `startObservingObjectsDidChangeNotifications()`.
+    ///
+    func startObservingStorageManagerDidResetNotifications() {
+        // Remove just in case this method was called already.
+        stopObservingStorageManagerDidResetNotifications()
+
+        storageManagerDidResetObservationToken =
+            notificationCenter.addObserver(forName: .StorageManagerDidResetStorage, object: storageManager, queue: nil) { _ in
```

Comment:
nit non-blocking suggestion: since this class is iOS 13+, we can take advantage of the [Combine way of observing a notification](https://developer.apple.com/documentation/combine/routing-notifications-to-combine-subscribers) 😄 then `stopObservingStorageManagerDidResetNotifications` could be just cancelling the `AnyCancellable` from the observation

Reply:
Great idea! Changed in 96000ead32790a98e49aadd85c8558ecf467ea9c and 711366647a6f4209789789dfdfec729da94ae611. I removed the `stop*` methods because I read that we don't need to also call `AnyCancellable.cancel()` in `deinit()`. Though I kept the `deinit` calls for now until I verify it myself. 😄